### DISCUSSION TITLE
Pagination with N+1 Issue

### DISCUSSION THREAD

**Discussion by jringeisen** [livewire/livewire#7966] [1. Help] [answered]
I'm stuck and hoping someone knows what's going on here. I've ran this across two other developers and they are just as confused as I am. I've implemented cursor pagination using Livewire 3 and I'm getting a lot of duplicate queries. The code snippet below works exactly as you expect. I get 24 items on initial page load, then each time I click load more 24 are added, but the queries grow exponentially.

Also, I created a fresh laravel/livewire application with one model called Post and it has a title attribute and I seeded 500 rows of data and I'm having the same N+1 problem so it's not a querying problem, it's something to do with how livewire is hydrating the data.

I just confirmed my suspension. I setup a fresh laravel/livewire 2.0 app and it doesn't do the N+1 problem. So either I'm not understanding what's going on in Livewire 3 or there is a bug.

<img width="632" alt="Screenshot 2024-02-16 at 9 50 04 PM" src="https://github.com/livewire/livewire/assets/7864804/d9e887c0-78c5-4562-aef1-a086ac6bf78d">

```php
<?php

namespace App\Http\Livewire;

use App\Models\Issue;
use Livewire\Component;
use Livewire\WithPagination;
use Illuminate\Pagination\Cursor;
use Illuminate\Support\Collection;
use Illuminate\Contracts\View\View;

class PublisherDetailPage extends Component
{
    use WithPagination;

    public Collection $issues;

    public int $perPage = 24;

    public ?bool $hasMorePages = null;
    public ?string $nextCursor = null;

    public function mount(): void
    {
        $this->issues = collect();
        $this->loadMore();
    }

    public function loadMore(): void
    {
        $moreIssues = Issue::query()
            ->cursorPaginate($this->perPage, ['*'], 'cursor', Cursor::fromEncoded($this->nextCursor));

        // When I comment out this snippet of code the N+1 problem goes away but obviously 
        // I need this to see the data on the page
        $this->issues = $this->issues->push(...$moreIssues->items());

        $this->hasMorePages = $moreIssues->hasMorePages();
        $this->nextCursor = $moreIssues->nextCursor()?->encode();
    }

    public function render(): View
    {
        return view('livewire.publisher-detail-page');
    }
}
```

**Comment by jringeisen** [comment, marked as answer]
This seems to have solved my issue. Lesson learned, don't use a property to assign the paginated values to. I think the way Livewire hydrates in version 3 causes issues when assigning paginated values to a property. Correct me if I'm wrong! Also, this works just fine without the `WithPagination` trait.

```php
<?php

namespace App\Livewire;

use App\Models\Post;
use Livewire\Component;
use Illuminate\Pagination\Cursor;
use Livewire\Attributes\Computed;

class Pagination extends Component
{
    public $perPage = 20;
    public $hasMorePages;
    public $nextCursor;

    public function loadMore()
    {
        $this->perPage += 20;
    }

    #[Computed()]
    public function paginate()
    {
        $posts = Post::cursorPaginate($this->perPage, ['*'], 'cursor', Cursor::fromEncoded($this->nextCursor));

        $this->hasMorePages = $posts->hasMorePages();
        $this->nextCursor = $posts->nextCursor()?->encode();

        return $posts->items();
    }

    public function render()
    {
        return view('livewire.pagination', [
            'posts' => $this->paginate(),
        ]);
    }
}
```

**Reply by joshhanley** [reply]
@jringeisen just saw this, and yep using a computed property is better than assigning collections to a public property.

But there are a couple of issues with your code, computed properties are called using `$this->paginate` instead of `$this->paginate()`. The way you are doing it currently, your just calling the method so you might as well not have the computed attribute.

Also computed properties are made available to the view, so you don't need to pass it in through the render method.

So you could refactor all of that to this:

```php
<?php

namespace App\Livewire;

use App\Models\Post;
use Livewire\Component;
use Livewire\WithPagination;
use Illuminate\Pagination\Cursor;
use Livewire\Attributes\Computed;

class Pagination extends Component
{
    use WithPagination;

    public $perPage = 20;
    public $hasMorePages;
    public $nextCursor;

    public function loadMore()
    {
        $this->perPage += 20;
    }

    #[Computed()]
    public function posts()
    {
        $posts = Post::cursorPaginate($this->perPage, ['*'], 'cursor', Cursor::fromEncoded($this->nextCursor));

        $this->hasMorePages = $posts->hasMorePages();
        $this->nextCursor = $posts->nextCursor()?->encode();

        return $posts->items();
    }

    public function render()
    {
        return view('livewire.pagination');
    }
}
```

I've renamed your computed property to `posts` so then in your view, you would call `$this->posts` to access it (note it needs the `$this->` to access a computed property. I've removed the posts being passed into view in render. And I've added `WithPagination` back.

You do need `WithPagination` if you want Livewire to handle the page change for you (aka without a full page refresh). Without it, the browser just goes to the new page but does a full page refresh in the process.

**Reply by joshhanley** [reply]
Regarding why the n+1 problem was happening, was because you were assigning a collection of models to a public property. 

Under the hood, the way Livewire keeps track of that, is it has to send the IDs of all the models in the collection and the class of model, to the frontend. On the next request, Livewire rehydrates that collection, by doing a DB query that has a `whereIn` and passes all the IDs in. So any custom selects, joins, relationships, etc. will be missing. Hence causing the n+1 issue.

That's why we recommend using computed properties over assigning collections to properties. See here for more details https://livewire.laravel.com/docs/properties#eloquent-constraints-arent-preserved-between-requests

Hope this helps!

**Reply by jringeisen** [reply]
@joshhanley that makes perfect sense. Thanks for answering, that's exactly what I needed. Now I understand it.

**Reply by joshhanley** [reply]
@jringeisen no worries! 🙂

**Reply by jringeisen** [reply]
@joshhanley, I have one final question. In Livewire 2, I used to paginate results and then append these paginated results to a property. Instead of increasing the perPage value on each loadMore invocation, it would consistently paginate a set number of results—let's say 10—for each loadMore. These results would then be added to a collection. Is this approach no longer feasible in Livewire 3?

**Reply by joshhanley** [reply]
@jringeisen in Livewire 2 that still would still have caused the collection to be restored from the database on every request (by Livewire doing it) plus you making a new query to get the extra ones. So in Livewire 3, you would be making a query in your computed property that selects all of the records up to the count amount. Which will actually be more efficient as it's only making one query per request.

Either way it's not an ideal set up for infinite pagination like that. Instead a better way is to make use of nested components. Nested components, once running on the page are isolated from all other components (including the parent).

So what you could do it keep track of the count and the "loadMore" button and method in the parent component, and for each "page" of results, create a new nested component which gets passed the limit and offset, and it handles making a database query only for the records it is showing. That way once they are on the page, they won't be touched when you click the loadMore button, keeping it as efficient as possible.

Hope that makes sense

**Reply by jringeisen** [reply]
Ahhhh I like the sound of that. I'm going to look into it. Thank you for the help!

**Reply by joshhanley** [reply]
No worries! 🙂

**Reply by rubendn** [reply]
Josh,
That sounds like a great suggestion on using another component to load the additional records.  How would you handle filters for the data that are in the parent livewire component.  Would you pass those as props to to the new nested component?  How would you clear the nested components when the filters change?

**Reply by joshhanley** [reply]
@rubendn yeah I would probably pass them down as props. Regarding the filter change, you should be resetting the pagination count, which will get rid of all the extra components.

But for the first one, I would probably add some of the filters to the `wire:key` or something like that, so when a filter changes, the key changes triggering Livewire to remove the old one and add a new one which will receive the new filters.

Or you could just generate a unique key that you store in the parent that the children use in their `wire:key` which you could regenerate the key if ever the filters change.
